### PR TITLE
[dtensor][8/N] Introduce cost model for sharding

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111278
* #111234
* #110717
* __->__ #109145

This PR adds some basic comm cost model for sharding prop

### CODE REVIEW

File: torch/distributed/_tensor/_collective_utils.py [hunk 1]
```diff
@@ -158,3 +160,139 @@ def mesh_all_to_all(
             async_op=async_op,
         )
     return work
+
+
+def spec_to_bytes(spec: "placement_types.DTensorSpec") -> int:
+    assert spec.tensor_meta is not None, "spec should have tensor meta defined!"
+    dtype_to_bytes = {
```
Comment (Chillee) [inline torch/distributed/_tensor/_collective_utils.py:167]: You can get `tensor.dtype.itemsize`

Comment (wanchaol) [reply torch/distributed/_tensor/_collective_utils.py:167]: cool good to know there's already a API for it, I'll update

File: torch/distributed/_tensor/_collective_utils.py [hunk 2]
```diff
@@ -158,3 +160,139 @@ def mesh_all_to_all(
...
+    # constant latency factor + bandwidth cost
+    return (
+        1
+        + bandwidth_factor
+        * num_bytes
+        * (num_devices_on_mesh_dim - 1)
+        / num_devices_on_mesh_dim
+    )
+
+
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_tensor/_collective_utils.py:244]: Here the costs are suitable for both intra node or inter-node or mixed, this is coming from the device mesh layout information where we know whether we can in the intra or inter node scenario

File: torch/distributed/_tensor/placement_types.py [hunk 3]
```diff
@@ -533,3 +533,9 @@ def from_dim_map(
                 placements[m] = Shard(i)
 
         return cls(mesh, tuple(placements), tensor_meta=tensor_meta)
+
+    def is_replicated(self):
```
Comment (fduwjj) [inline torch/distributed/_tensor/placement_types.py:537]: nice!

File: torch/distributed/_tensor/_collective_utils.py [hunk 4]
```diff
@@ -158,3 +160,128 @@ def mesh_all_to_all(
             async_op=async_op,
         )
     return work
+
+
+def spec_to_bytes(spec: "placement_types.DTensorSpec") -> int:
```
Comment (fduwjj) [inline torch/distributed/_tensor/_collective_utils.py:165]: To clarify, the result is in  bytes right?

File: torch/distributed/_tensor/_collective_utils.py [hunk 5]
```diff
@@ -158,3 +160,128 @@ def mesh_all_to_all(
...
+    assert spec.tensor_meta is not None, "spec should have tensor meta defined!"
+    return spec.tensor_meta.dtype.itemsize * math.prod(spec.shape)
+
+
+def get_bandwidth_factor(mesh: DeviceMesh) -> List[float]:
+    # generate bandwidth factor for intra-host/inter-host communication pattern
+    factors = [1.0 * mesh.ndim]
```
Comment (fduwjj) [inline torch/distributed/_tensor/_collective_utils.py:172]: Does this mean we are generating a list of one item or a list of mesh.ndim 1.0s?, IIUC, we want latter?

**Review by fduwjj — COMMENTED** [review]
Do we want to add at least one simple UT for the cost cost modeling?

File: torch/distributed/_tensor/_collective_utils.py (hunk 4 above)
Comment (wanchaol) [reply torch/distributed/_tensor/_collective_utils.py:165]: yep it's #num bytes

File: torch/distributed/_tensor/_collective_utils.py (hunk 5 above)
Comment (wanchaol) [reply torch/distributed/_tensor/_collective_utils.py:172]: good catch!!

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 6]
```diff
@@ -121,6 +121,7 @@ def gen_bucketize_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyTyp
     Just propagate input sharding, but expect replicated for boundaries input.
     """
     input_strategy = op_schema.args_schema[0]
+    boundaries_strategy = op_schema.args_schema[1]
```
Comment (fduwjj) [inline torch/distributed/_tensor/ops/tensor_ops.py:124]: Do we need this one?

**Review by fduwjj — APPROVED** [review]
LGTM

**Pull request closed without merging**